src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -1431,5 +1456,208 @@
+            Debug.Assert(testBothArgs == testBothArgs0);
+
+            var right = VisitExpression(binOp.Right);
+            if (forceSpill || ShouldSpill(right))
+            {
+                var left = VisitExpression(binOp.Left);
+                dest ??= _F.GenerateLabel("dest");
+                builder ??= new BoundSpillSequenceBuilder(binOp.Syntax);
+                if (testBothArgs)
+                {
+                    LabelSymbol? fallThrough = null;
+                    SpillCondBranch(builder, left, ref fallThrough, !sense);
+                    SpillCondBranch(builder, right, ref dest, sense);
+                    if (fallThrough is not null)
+                    {
+                        builder.AddStatement(_F.Label(fallThrough));
+                    }
+                }
+                else
+                {
+                    SpillCondBranch(builder, left, ref dest, sense);
+                    SpillCondBranch(builder, right, ref dest, sense);
+                }
+
+                return null;
+            }
+            else
+            {
+                right = ConsolidateSpillNode(ref builder, right, topLevel: false);
+
+                BoundExpression left;
+                if (builder is null)
+                {
+                    left = VisitExpression(ref builder, binOp.Left, topLevel);
+                }
+                else
+                {
+                    var leftBuilder = new BoundSpillSequenceBuilder(builder.Syntax);
+                    left = VisitExpression(ref leftBuilder, binOp.Left, topLevel);
+                    leftBuilder.Include(builder);
+                    builder = leftBuilder;
+                }
+
+                return UpdateExpression(builder, binOp.Update(
+                    binOp.OperatorKind, binOp.ConstantValueOpt,
+                    binOp.Method, binOp.ConstrainedToType, binOp.ResultKind,
+                    left: left, right: right, binOp.Type));
+            }
+        }
+
+        private void SpillIsPatternCondExpression(BoundSpillSequenceBuilder builder, BoundLoweredIsPatternExpression node, LabelSymbol dest, bool sense)
+        {
+            _F.Syntax = node.Syntax;
+
+            builder.AddLocals(node.Locals);
+            builder.AddStatements(node.Statements);
+            if (sense)
+            {
+                builder.AddStatement(_F.Label(node.WhenTrueLabel));
+                builder.AddStatement(_F.Goto(dest));
+                builder.AddStatement(_F.Label(node.WhenFalseLabel));
+            }
+            else
+            {
+                var doneLabel = _F.GenerateLabel("doneLabel");
+                builder.AddStatement(_F.Label(node.WhenTrueLabel));
+                builder.AddStatement(_F.Goto(doneLabel));
+                builder.AddStatement(_F.Label(node.WhenFalseLabel));
+                builder.AddStatement(_F.Goto(dest));
+                builder.AddStatement(_F.Label(doneLabel));
+            }
+        }
+
```

Comment:
>SpillIsPatternExpression [](http://example.com/codeflow?start=27&length=24)

I hope we could be smarter about how we spill BoundLoweredIsPatternExpression and avoid introducing result locals in the process. I think BoundLoweredIsPatternExpression node is restricted enough in terms of what it contains, effectively that is just a complex expression:
- It represents a value
- It has one input value and it has one output value. 
- there is no control flow out of the node (no `yield  return`, no awaits, no gotos out of the node)
- The input value is evaluated first
- The other parts that are evaluated are supposed to have no observable side-effects on the outside world. Therefore, evaluation of them  can be performed "out of order", at a later time when we actually nee the output value.

Bottom line, all we need to worry about is to evaluate the input value at the right time (in terms of order of evaluation). We do not have to calculate the result of the pattern at the same time though.

Let's say an await triggers a condition when we need to spill the value represented by BoundLoweredIsPatternExpression. I think all we need to do is to spill the input value and then reassemble the BoundLoweredIsPatternExpression back and return it. The emit layer will do the rest.

Reply:
> Therefore, evaluation of them can be performed "out of order", at a later time when we actually need the output value.

On the second thought, I am probably wrong about that since the state of the input object could be changed by expressions that are evaluated after the pattern. In the next argument, for example.

So, the current spilling is probably fine.

Reply:
Removed entirely, by default it'll use an assignment to a synthesized local which is fine for this.

Reply:
Removed entirely, by default it'll use an assignment to a synthesized local which is fine for this.

Reply:
> The evaluation of the input expression happens in inner statements (we don't have a separate field for the input to spill it directly)

Why can't we spill the entire result as the value? We don't really need to spill the input expression separately. What am I missing.  

> spiller won't be triggered for those anyways since we expect an empty stack there.

I am not following. What is the scenario? If spiller won't be triggered, why do we need to spill?

Reply:
>Why can't we spill the entire result as the value? 

Ok, I tried that but the optimizer will throw on missing local key (even if moved to a Sequence in case I was missing something), I couldn't figure what's the issue but it's probably due this invariant of empty stack for statements - locals are declared at current stack. I'll have to investigate more but I'd agree that we should be able to simply spill as a value.

I think we need to deal with this stack expectation at every step post local rewrite which is not exactly easy to spot.

> We don't really need to spill the input expression separately.

What I mean is that for any expression we visit sub-expressions to see if spilling is required. Right now, is-expressions are always spilled (for a different reason - using SpillSequence) but still we can't know for sure because the only thing we see is a list of statements.

Reply:
> Right now, is-expressions are always spilled (for a different reason - using SpillSequence

I thought we stopped creating SpillSequence for is pattern. What am I missing?

Reply:
(Right now, as in the current version of the compiler)